src/action.ts

```diff
@@ -0,0 +1,117 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Action implementation.
+ */
+import { error, info, debug } from '@actions/core';
+
+import { Inputs } from './inputs';
+import { spawn } from 'child_process';
+import ms from 'milliseconds';
+
+const OS = process.platform;
+
+function getExecutable(): string {
+  return OS === 'win32' ? 'powershell' : 'bash';
+}
+
+async function wait(ms: number) {
+  return new Promise((r) => setTimeout(r, ms));
+}
+
+interface CommandResult {
+  success: boolean;
+  exitCode: number;
+  output: string[];
+}
+
+async function runCommand(inputs: Inputs): Promise<CommandResult> {
+  const endTime = Date.now() + ms.hours(5);
+  const executable = getExecutable();
+
+  // Timeout exit code - 124
+  let exitCode = 124;
+  let done = false;
+  const output: string[] = [];
+
+  debug(`Running command ${inputs.command} on ${OS} using shell ${executable}`);
+  const child = spawn(inputs.command, { shell: executable });
+
+  child.stdout?.on('data', (data) => {
+    process.stdout.write(data);
+    output.push(data);
+  });
+  child.stderr?.on('data', (data) => {
+    process.stdout.write(data);
+    output.push(data);
+  });
+
+  child.on('exit', (code) => {
+    debug(`Code: ${code}`);
+
+    if (code === null) {
+      error('exit code cannot be null');
+      exitCode = -1;
+      return;
+    }
+
+    exitCode = code;
+    done = true;
+  });
+
+  do {
+    const pollingPeriod = ms.seconds(1);
+    await wait(pollingPeriod);
+  } while (Date.now() < endTime && !done);
+
+  return {
+    success: exitCode === 0,
+    exitCode,
+    output,
+  };
+}
+
+function hasFlakyOutput(flaky_test_output_lines: string[], output: string[]): boolean {
+  const flakyIndicator = flaky_test_output_lines.find((flakyLine) =>
+    output.some((outputLine) => outputLine.includes(flakyLine))
+  );
+  if (flakyIndicator === undefined) {
+    info("Output doesn't contain flaky lines");
+    return false;
+  }
+  info(`Output contains flaky line: ${flakyIndicator}`);
+  return true;
+}
+
+export async function runAction(inputs: Inputs): Promise<number> {
+  for (let attempt = 1; attempt <= inputs.maxAttempts; attempt++) {
+    info(`Starting attempt #${attempt}`);
+    const { success, exitCode, output } = await runCommand(inputs);
+    if (success) {
+      info(`Attempt #${attempt} succeeded`);
+      return 0;
+    }
+
+    info(`Attempt #${attempt} failed with exit code ${exitCode}`);
+    if (attempt == inputs.maxAttempts) {
+      return exitCode;
+    }
+
+    if (!hasFlakyOutput(inputs.flakyTestOutputLines, output)) {
+      return exitCode;
+    }
```

Comment:
I don't mind, but note that we already will have
"Output contains flaky line: ${flakyIndicator}" log + "Starting attempt #${attempt}". IMO it's enough, but I can add if you want

Reply:
Ah thanks, I missed that, sorry -- but I think the call to info() is in the wrong place. Can we move it to this function?

Based on its name, hasFlakyOutput() is something that would be expected to just return a boolean and shouldn't have any meaningful side effects, whereas we do want printing the log to be a deliberate action that we take in the execution.

I also suggest adding "Restarting test" or similar to that message since it connects the restart with the error.

Reply:
> Based on its name, hasFlakyOutput() is something that would be expected to just return a boolean and shouldn't have any meaningful side effects, whereas we do want printing the log to be a deliberate action that we take in the execution.

I agree with that, but I think it's very helpful to output which flake indicator was found, so we need to output something in the hasFlakyOutput method.

I moved logging to the calling side, and also left the log of what indicator was found.